### ISSUE TITLE
Parser cache test shows massive performance degradation

### ISSUE THREAD

**Issue by beikov** [Blazebit/blaze-persistence#228]
Before the changes, the total time was at about 6 seconds
https://travis-ci.org/Blazebit/blaze-persistence/jobs/151804269#L432

Now it's at 32 seconds
https://travis-ci.org/Blazebit/blaze-persistence/jobs/158969281#L445

Investigate what happened and fix it please.

**Comment by Mobe91** [comment]
The issue was introduced in af0ee2e with the macros. I still do not understand why my fix works since I merely changed

```
MACRO: { macros != null && tryMacro() }? .;
```
to

```
MACRO: . { macros != null && tryMacro() }?;
```

I created a question on Stackoverflow for this:
http://stackoverflow.com/questions/39493251/semantic-lexer-predicate-performance

**Comment by Mobe91** [comment]
The "fix" in c88e2a6 was actually wrong since the altered rule failed to accept the macro. The performance increase was just because the semantic predicate was evaluated just once.
